src/mpi/init/init_async.c

```diff
@@ -59,22 +73,117 @@ static void progress_fn(void *data)
+    char *proc_id_str, *tmp;
+    size_t proc_count;
+
+    if (!affinity_to_parse || strlen(affinity_to_parse) == 0) {
+        /* generate default affinity */
+        proc_count = MPL_get_nprocs();
+        for (th_idx = 0; th_idx < threads_per_node; th_idx++) {
+            if (th_idx < proc_count)
+                thread_affinity[th_idx] = proc_count - th_idx - 1;
+            else
+                thread_affinity[th_idx] = thread_affinity[th_idx % proc_count];
+        }
+        goto fn_exit;
+    }
+
+    /* create copy of original buffer because it will be modified in strsep */
+    affinity_copy = MPL_strdup(affinity_to_parse);
+    MPIR_Assert(affinity_copy);
+    tmp = affinity_copy;
+
+    for (th_idx = 0; th_idx < threads_per_node; th_idx++) {
+        proc_id_str = MPL_strsep(&tmp, " \n\t,");
+        if (proc_id_str != NULL) {
+            if (atoi(proc_id_str) < 0) {
+                printf
+                    ("ERROR: unexpected proc_id %s, affinity string %s\n",
+                     proc_id_str, affinity_to_parse);
+                mpi_errno = MPI_ERR_OTHER;
+                goto fn_fail;
+            }
+            thread_affinity[th_idx] = atoi(proc_id_str);
+            read_count++;
+        } else {
+            printf
+                ("ERROR: unexpected end of affinity string, expected %d numbers, read %d, affinity string %s\n",
+                 threads_per_node, read_count, affinity_to_parse);
+            mpi_errno = MPI_ERR_OTHER;
+            goto fn_fail;
+        }
+    }
+    if (read_count < threads_per_node) {
+        printf
+            ("ERROR: unexpected number of processors (specify 1 logical processor per 1 progress thread), affinity string %s\n",
+             affinity_to_parse);
+        mpi_errno = MPI_ERR_OTHER;
+        goto fn_fail;
+    }
+
+  fn_exit:
+    if (MPIR_Process.comm_world->rank == 0) {
+        for (th_idx = 0; th_idx < threads_per_node; th_idx++) {
+            MPL_DBG_MSG_FMT(MPIDI_CH4_DBG_GENERAL, VERBOSE,
+                            (MPL_DBG_FDEST, "affinity: thread %d, processor %d",
+                             th_idx, thread_affinity[th_idx]));
+        }
+    }
+    MPL_free(affinity_copy);
+    return mpi_errno;
+
+  fn_fail:
+    goto fn_exit;
+}
+
 /* called inside MPID_Init_async_thread to provide device override */
 int MPIR_Init_async_thread(void)
 {
-    int mpi_errno = MPI_SUCCESS;
+    int mpi_errno = MPI_SUCCESS, thr_err;
+    int global_rank, local_rank, local_size, threads_per_node;
+    int *thread_affinity = NULL;
     MPIR_FUNC_TERSE_STATE_DECL(MPID_STATE_MPIR_INIT_ASYNC_THREAD);
 
     MPIR_FUNC_TERSE_ENTER(MPID_STATE_MPIR_INIT_ASYNC_THREAD);
 
+    global_rank = MPIR_Process.comm_world->rank;
+    local_rank =
+        (MPIR_Process.comm_world->node_comm) ? MPIR_Process.comm_world->node_comm->rank : 0;
+    local_size =
+        (MPIR_Process.comm_world->node_comm) ? MPIR_Process.comm_world->node_comm->local_size : 1;
```

Comment:
You can directly use `MPIR_Process.local_rank` and `MPIR_Process.local_size`.

Reply:
If we use them, would they still be equivalent to the following? 
```
        local_rank =
            (MPIR_Process.comm_world->node_comm) ? MPIR_Process.comm_world->node_comm->rank : 0;
```
```
        if (have_cliques) {
            local_size = MPIR_Process.comm_world->local_size;
        } else {
            local_size =
                (MPIR_Process.comm_world->node_comm) ? MPIR_Process.comm_world->
                node_comm->local_size : 1;
        }
```